vesc_hw_interface/src/vesc_hw_interface.cpp

```diff
@@ -407,28 +427,102 @@ void VescHwInterface::packetCallback(const std::shared_ptr<VescPacket const>& pa
+  {
+    std::shared_ptr<VescPacketMCConf const> mc_conf = std::dynamic_pointer_cast<VescPacketMCConf const>(packet);
+
+    auto config = mc_conf->getConfig();
+    num_rotor_poles_ = config.si_motor_poles;
+    gear_ratio_ = config.si_gear_ratio;
+    if (config.motor_type == MOTOR_TYPE_FOC) {
+      auto pole_pairs = num_rotor_poles_ / 2.0;
+      auto flux_linkage = config.foc_motor_flux_linkage;
+      torque_const_ = (60.0 / (2.0 * M_PI * pole_pairs)) * flux_linkage;
+    }
+    RCLCPP_INFO_STREAM(rclcpp::get_logger("VescHwInterface"), "Extracted configuration from VESC:");
+    RCLCPP_INFO_STREAM(rclcpp::get_logger("VescHwInterface"), "  - Number of rotor poles: " << num_rotor_poles_);
+    RCLCPP_INFO_STREAM(rclcpp::get_logger("VescHwInterface"), "  - Gear ratio: " << gear_ratio_);
+    RCLCPP_INFO_STREAM(rclcpp::get_logger("VescHwInterface"), "  - Torque constant: " << torque_const_);
   }
   else if (packet->getName() == "Values")
   {
     std::shared_ptr<VescPacketValues const> values = std::dynamic_pointer_cast<VescPacketValues const>(packet);
 
-    const double current = values->getMotorCurrent();
-    const double velocity_rpm = values->getVelocityERPM() / static_cast<double>(num_rotor_poles_ / 2);
-    const double steps = values->getPosition();
+    const auto current = values->getMotorCurrent();
+    const auto velocity_rpm = values->getVelocityERPM() / static_cast<double>(num_rotor_poles_ / 2);
+    const auto position = values->getPosition();
+    const auto steps = static_cast<int32_t>(values->getTachometer());
 
-    position_ = steps / (num_hall_sensors_ * num_rotor_poles_) * gear_ratio_;  // unit: rad or m
-    velocity_ = velocity_rpm / 60.0 * 2.0 * M_PI * gear_ratio_;                // unit: rad/s or m/s
-    effort_ = current * torque_const_ / gear_ratio_;                           // unit: Nm or N
-  }
+    if (!homing_done_ && homing_enabled_)
+    {
+      servo_controller_.updateSensor(packet);
+      homing_offset_ = position;
+      return;
+    }
+    if (!sensor_initialize_)
+    {
+      if (joint_type_ == "revolute" || joint_type_ == "prismatic")
+      {
+        if (!homing_enabled_)
+        {
+          sensor_initialize_ = true;
+          return;
+        }
+        sensor_initialize_ = (std::fabs(homing_offset_ - position) < std::numeric_limits<double>::epsilon()) ? true : false;
+        homing_offset_ = position;
+      }
+      else if (joint_type_ == "continuous")
+      {
+        sensor_initialize_ = (steps == prev_steps_) ? true : false;
+        prev_steps_ = steps;
+      }
+      RCLCPP_INFO_STREAM(rclcpp::get_logger("VescHwInterface"), "waiting for values to settle...");
+      return;
+    }
 
-  return;
+    // calculate position
+    if (joint_type_ == "revolute" || joint_type_ == "prismatic")
+    {
+      // `position` is [deg] but here we mapped the position to the joint limits hence unit is irrelevant
+      position_ = std::fmod(position - homing_offset_ + 360.0, 360.0);
```

Comment:
I think it would be better if values like 360.0, 90.0, and 225.0 were expressed in the form of `const_var / n` or something similar.